library/psa_crypto_pake.c

```diff
@@ -0,0 +1,720 @@
+    PSA_PAKE_STEP_X2S           = 2,
+    PSA_PAKE_STEP_DERIVE        = 3,
+};
+
+enum psa_pake_state
+{
+    PSA_PAKE_STATE_INVALID      = 0,
+    PSA_PAKE_STATE_SETUP        = 1,
+    PSA_PAKE_STATE_READY        = 2,
+    PSA_PAKE_OUTPUT_X1_X2       = 3,
+    PSA_PAKE_OUTPUT_X2S         = 4,
+    PSA_PAKE_INPUT_X1_X2        = 5,
+    PSA_PAKE_INPUT_X4S          = 6,
+};
+
+enum psa_pake_sequence
+{
+    PSA_PAKE_SEQ_INVALID        = 0,
+    PSA_PAKE_X1_STEP_KEY_SHARE  = 1,    /* also X2S & X4S KEY_SHARE */
+    PSA_PAKE_X1_STEP_ZK_PUBLIC  = 2,    /* also X2S & X4S ZK_PUBLIC */
+    PSA_PAKE_X1_STEP_ZK_PROOF   = 3,    /* also X2S & X4S ZK_PROOF */
+    PSA_PAKE_X2_STEP_KEY_SHARE  = 4,
+    PSA_PAKE_X2_STEP_ZK_PUBLIC  = 5,
+    PSA_PAKE_X2_STEP_ZK_PROOF   = 6,
+    PSA_PAKE_SEQ_END            = 7,
+};
+
+#if defined(MBEDTLS_PSA_BUILTIN_PAKE)
+psa_status_t psa_pake_setup( psa_pake_operation_t *operation,
+                             const psa_pake_cipher_suite_t *cipher_suite)
+{
+    /* A context must be freshly initialized before it can be set up. */
+    if( operation->alg != 0 || operation->state != PSA_PAKE_STATE_INVALID )
+    {
+        return( PSA_ERROR_BAD_STATE );
+    }
+
+    if( cipher_suite == NULL ||
+        PSA_ALG_IS_PAKE(cipher_suite->algorithm ) == 0 ||
+        ( cipher_suite->type != PSA_PAKE_PRIMITIVE_TYPE_ECC &&
+          cipher_suite->type != PSA_PAKE_PRIMITIVE_TYPE_DH ) ||
+        PSA_ALG_IS_HASH( cipher_suite->hash ) == 0 )
+    {
+        return( PSA_ERROR_INVALID_ARGUMENT );
+    }
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_ECJPAKE)
+    if( cipher_suite->algorithm != PSA_ALG_JPAKE ||
+        cipher_suite->type != PSA_PAKE_PRIMITIVE_TYPE_ECC ||
+        cipher_suite->family != PSA_ECC_FAMILY_SECP_R1 ||
+        cipher_suite->bits != 256 ||
+        cipher_suite->hash != PSA_ALG_SHA_256 )
+    {
+        return( PSA_ERROR_NOT_SUPPORTED );
+    }
+
+    operation->alg = cipher_suite->algorithm;
+
+    mbedtls_ecjpake_init( &operation->ctx.ecjpake );
+
+    operation->state = PSA_PAKE_STATE_SETUP;
+    operation->sequence = PSA_PAKE_SEQ_INVALID;
+    operation->input_step = PSA_PAKE_STEP_X1_X2;
+    operation->output_step = PSA_PAKE_STEP_X1_X2;
+
+    operation->buffer = NULL;
+    operation->buffer_length = 0;
+    operation->buffer_offset = 0;
+
+    return( PSA_SUCCESS );
+#else
+    return( PSA_ERROR_NOT_SUPPORTED );
+#endif
+}
+
+psa_status_t psa_pake_set_password_key( psa_pake_operation_t *operation,
+                                        mbedtls_svc_key_id_t password )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    psa_key_attributes_t attributes = psa_key_attributes_init();
+    psa_key_type_t type;
+    psa_key_usage_t usage;
+
+    if( operation->alg == 0 ||
+        operation->state != PSA_PAKE_STATE_SETUP )
+    {
+        return( PSA_ERROR_BAD_STATE );
+    }
+
+    status = psa_get_key_attributes( password, &attributes );
+    if( status != PSA_SUCCESS )
+        return status;
```

Comment:
Style: half of the return statements in this function have braces, half of them don't. This is a new file, so it should at least start in a consistent state.

(Sorry, I know this part of our coding style is stupi^unusual and goes against finger memory for most C coders. Again, I can't wait for us to change it, but it hasn't happened yet.)

Reply:
I'm ready to do a full cleanup of the braces, can we ack on the rules :
- single-line condition: no braces
- multi-line condition: braces

What about braces in a single line code in an else after braces:
```
if( condition )
{
    code;
    code;
}
```

Should be:
```
else
   code
```
or
```
else
{
    else
}
````

For instance the Linux code style asks to keep brace usage the same on all conditions, meaning if the main if has braces, the else would have brace even if there's a single line of code. And if the first condition has a single line of code, but the else multiple lines, braces should be added on the first condition. (see [linux coding style](https://www.kernel.org/doc/html/latest/process/coding-style.html#placing-braces-and-spaces)).

It would help to have those clarified.

Reply:
Ah, sorry, I miswrote, I meant parentheses here: have of the return statements look like `return PSA_xxx;` and the other half like `return( PSA_xxx );`. About this point, I'm afraid our current coding style is unambiguous: we're supposed to have parentheses (even though it's highly non-standard and I'm not a big fan).

Regarding braces for the else branch, I don't think our current coding style is precise enough to specify that, but I think we should have braces on all branches as soon as one of the branches body, or the initial condition, is more than one line (similarly to the Linux style).

Reply:
Our coding style guide doesn't say, our current usage is inconsistent, and we're going to switch to a new style Soon™. So it doesn't matter.

Reply:
@mpg oh ok, indeed I'll fix the return statements
@gilles-peskine-arm ok I'll let braces structure as-is then